File: Project/NetworkingManaged/BaseSocket.cs
﻿// Copyright 2019. All Rights Reserved.
using GameFramework.BinarySerializer;
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using System.Threading;

namespace GameFramework.Networking
{
	public abstract class BaseSocket
	{
		protected abstract class EventBase
		{
			public EventBase()
			{
			}
		}

		private class EventBaseList : List<EventBase>
		{ }

		protected class SendCommand
		{
			public BufferStream Buffer
			{
				get;
				private set;
			}

			public double SendTime
			{
				get;
				private set;
			}

			public SendCommand(BufferStream Buffer, double SendTime)
			{
				this.Buffer = Buffer;
				this.SendTime = SendTime;
			}
		}

		private class SendCommandList : List<SendCommand>
		{ }

		private Thread receiveThread = null;
		private Thread sendThread = null;

		private EventBaseList events = null;
		private SendCommandList sendCommands = null;

		public Protocols Type
		{
			get;
			private set;
		}

		protected Socket Socket
		{
			get;
			private set;
		}

		protected byte[] ReceiveBuffer
		{
			get;
			private set;
		}

		protected uint ReceiveBufferIndex
		{
			get;
			set;
		}

		public abstract bool IsReady
		{
			get;
		}

		public abstract double Timestamp
		{
			get;
		}

		public NetworkingStatistics Statistics
		{
			get;
			private set;
		}

		public uint SendBufferSize
		{
			get;
			set;
		}

		public uint ReceiveBufferSize
		{
			get;
			set;
		}

		public bool MultithreadedCallbacks
		{
			get;
			set;
		}

		public bool MultithreadedReceive
		{
			get;
			set;
		}

		public bool MultithreadedSend
		{
			get;
			set;
		}

		public float PacketLossSimulation
		{
			get;
			set;
		}

		public int LatencySimulation
		{
			get;
			set;
		}

		public BaseSocket(Protocols Type)
		{
			this.Type = Type;

			events = new EventBaseList();
			sendCommands = new SendCommandList();

			SendBufferSize = Constants.SEND_BUFFER_SIZE;
			ReceiveBufferSize = Constants.RECEIVE_BUFFER_SIZE;

			Socket = SocketUtilities.CreateSocket(Type);
			SocketUtilities.SetBlocking(Socket, false);
			SocketUtilities.SetReceiveTimeout(Socket, Constants.RECEIVE_TIMEOUT);
			SocketUtilities.SetSendTimeout(Socket, Constants.SEND_TIMEOUT);
			SocketUtilities.SetTimeToLive(Socket, Constants.TIME_TO_LIVE);
			SocketUtilities.SetIPv6OnlyEnabled(Socket, false);
			//SocketUtilities.SetChecksumEnabled(Socket, false);

			if (Type == Protocols.TCP)
				SocketUtilities.SetNagleAlgorithmEnabled(Socket, false);

			Statistics = new NetworkingStatistics();

			MultithreadedCallbacks = true;
			MultithreadedReceive = true;
			MultithreadedSend = true;
		}

		public virtual void Service()
		{
			if (!MultithreadedReceive && IsReady)
				Receive();

			if (!MultithreadedSend && IsReady)
				HandleSendCommands();

			if (!MultithreadedCallbacks)
			{
				lock (events)
				{
					for (int i = 0; i < events.Count; ++i)
						ProcessEvent(events[i]);

					for (int i = 0; i < events.Count; ++i)
						events.RemoveAt(0);
				}
			}
		}

		protected void Shutdown()
		{
			if (MultithreadedReceive)
				receiveThread.Abort();

			if (MultithreadedSend)
				sendThread.Abort();

			SocketUtilities.CloseSocket(Socket);
		}

		protected void RunReceiveThread()
		{
			SocketUtilities.SetSendBufferSize(Socket, SendBufferSize);
			SocketUtilities.SetReceiveBufferSize(Socket, ReceiveBufferSize);
			ReceiveBuffer = new byte[ReceiveBufferSize];

			if (!MultithreadedReceive)
				return;

			receiveThread = new Thread(ReceiverWorker);
			receiveThread.Start();
		}

		protected void RunSenndThread()
		{
			if (!MultithreadedSend)
				return;

			sendThread = new Thread(SendWorker);
			sendThread.Start();
		}

		protected abstract void Receive();

		protected virtual bool SendOverSocket(Socket Target, BufferStream Buffer)
		{
			try
			{
				lock (Target)
				{
					if (PacketLossSimulation == 0 || Constants.Random.NextDouble() > PacketLossSimulation)
						Target.Send(Buffer.Buffer);

					Statistics.AddBandwidthOut(Buffer.Size);
				}

				return true;
			}
			catch (SocketException e)
			{
				if (e.SocketErrorCode == SocketError.ConnectionReset ||
					e.SocketErrorCode == SocketError.ConnectionAborted ||
					e.SocketErrorCode == SocketError.ConnectionRefused)
				{
					HandleDisconnection(Target);

					return false;
				}

				throw e;
			}
			catch (ObjectDisposedException e)
			{
				HandleDisconnection(Target);

				return false;
			}
			catch (Exception e)
			{
				throw e;
			}
		}

		protected virtual bool SendOverSocket(IPEndPoint EndPoint, BufferStream Buffer)
		{
			try
			{
				if (PacketLossSimulation == 0 || Constants.Random.NextDouble() > PacketLossSimulation)
					Socket.SendTo(Buffer.Buffer, EndPoint);

				Statistics.AddBandwidthOut(Buffer.Size);

				return true;
			}
			catch (SocketException e)
			{
				if (e.SocketErrorCode == SocketError.ConnectionReset ||
					e.SocketErrorCode == SocketError.ConnectionAborted ||
					e.SocketErrorCode == SocketError.ConnectionRefused)
				{
					return false;
				}

				throw e;
			}
			catch (Exception e)
			{
				throw e;
			}
		}

		protected abstract bool HandleSendCommand(SendCommand Command);

		protected abstract void ProcessEvent(EventBase Event);

		protected void AddEvent(EventBase Event)
		{
			lock (events)
				events.Add(Event);
		}

		protected void AddSendCommand(SendCommand Command)
		{
			lock (sendCommands)
				sendCommands.Add(Command);
		}

		protected virtual void HandleDisconnection(Socket Socket)
		{
			//Internationally Left Blank
		}

		private void ReceiverWorker()
		{
			while (true)
			{
				Thread.Sleep(1);

				Receive();
			}
		}

		private void SendWorker()
		{
			while (true)
			{
				Thread.Sleep(1);

				HandleSendCommands();
			}
		}

		private void HandleSendCommands()
		{
			lock (sendCommands)
			{
				for (int i = 0; i < sendCommands.Count; ++i)
				{
					SendCommand command = sendCommands[i];

					if (Timestamp < command.SendTime + (LatencySimulation / 1000.0F))
						continue;

					//if (!HandleSendCommand(sendCommands[i]))
					//	continue;
					HandleSendCommand(sendCommands[i]);

					sendCommands.RemoveAt(i--);
				}
			}
		}
	}
}

File: Project/NetworkingNative/Compile/BaseSocket.cpp
// Copyright 2019. All Rights Reserved.
#include "..\Include\BaseSocket.h"
#include "..\Include\Constants.h"

namespace GameFramework::Networking
{
	void GlobalReceiverWorker(BaseSocket* Instance)
	{
		Instance->ReceiverWorker();
	}

	void GlobalSendWorker(BaseSocket* Instance)
	{
		Instance->SendWorker();
	}

	BaseSocket::BaseSocket(PlatformNetwork::IPProtocols Type) :
		m_ReceiveThread(nullptr),
		m_SendThread(nullptr),
		m_EventsLock(0),
		m_Type(Type),
		m_Socket(0),
		m_ReceiveBuffer(nullptr),
		m_ReceiveBufferIndex(0),
		m_SendBufferSize(Constants::SEND_BUFFER_SIZE),
		m_ReceiveBufferSize(Constants::RECEIVE_BUFFER_SIZE),
		m_MultithreadedCallbacks(true),
		m_MultithreadedReceive(true),
		m_MultithreadedSend(true),
		m_PacketLossSimulation(0),
		m_LatencySimulation(0)
	{
		m_Socket = SocketUtilities::CreateSocket(m_Type);
		SocketUtilities::SetBlocking(m_Socket, false);
		SocketUtilities::SetReceiveBufferSize(m_Socket, m_ReceiveBufferSize);
		SocketUtilities::SetSendBufferSize(m_Socket, m_SendBufferSize);
		SocketUtilities::SetReceiveTimeout(m_Socket, Constants::RECEIVE_TIMEOUT);
		SocketUtilities::SetSendTimeout(m_Socket, Constants::SEND_TIMEOUT);
		SocketUtilities::SetTimeToLive(m_Socket, Constants::TIME_TO_LIVE);
		SocketUtilities::SetIPv6OnlyEnabled(m_Socket, false);
		//SocketUtilities::SetChecksumEnabled(m_Socket, false);

		if (m_Type == PlatformNetwork::IPProtocols::TCP)
			SocketUtilities::SetNagleAlgorithmEnabled(m_Socket, false);

		m_ReceiveBuffer = new std::byte[m_ReceiveBufferSize];

		m_MultithreadedCallbacks = true;
		m_MultithreadedReceive = true;
		m_MultithreadedSend = true;
	}

	BaseSocket::~BaseSocket(void)
	{
		Shutdown();
	}

	void BaseSocket::Service(void)
	{
		if (!m_MultithreadedReceive && GetIsReady())
			Receive();

		if (!m_MultithreadedSend && GetIsReady())
			HandleSendCommands();

		if (!m_MultithreadedCallbacks)
		{
			WAIT_FOR_BOOL(m_EventsLock);

			for (uint32_t i = 0; i < m_Events.size(); ++i)
			{
				EventBase* ev = m_Events[i];

				ProcessEvent(ev);

				delete ev;
			}

			for (uint32_t i = 0; i < m_Events.size(); ++i)
				m_Events.removeAt(0);
		}
	}

	void BaseSocket::Shutdown(void)
	{
		SocketUtilities::CloseSocket(m_Socket);

		if (m_MultithreadedReceive)
		{
			m_ReceiveThread->detach();
			delete m_ReceiveThread;
		}

		if (m_MultithreadedSend)
		{
			m_SendThread->detach();
			delete m_SendThread;
		}

		delete[]m_ReceiveBuffer;
	}

	void BaseSocket::RunReceiveThread(void)
	{
		if (!m_MultithreadedReceive)
			return;

		m_ReceiveThread = new thread(GlobalReceiverWorker, this);
	}

	void BaseSocket::RunSenndThread(void)
	{
		if (!m_MultithreadedSend)
			return;

		m_SendThread = new thread(GlobalSendWorker, this);
	}

	bool BaseSocket::SendOverSocket(Socket Target, const BufferStream& Buffer)
	{
		try
		{
			if (m_PacketLossSimulation == 0 || Constants::Random.NextDouble() > m_PacketLossSimulation)
				SocketUtilities::Send(Target, Buffer.GetBuffer(), Buffer.GetSize());

			m_Statistics.AddBandwidthOut(Buffer.GetSize());

			return true;
		}
		catch (PlatformNetwork::SocketException e)
		{
			if (e.GetError() == PlatformNetwork::Errors::ConnectionReset ||
				e.GetError() == PlatformNetwork::Errors::ConnectionAborted ||
				e.GetError() == PlatformNetwork::Errors::ConnectionRefused)
			{
				HandleDisconnection(Target);

				return false;
			}

			throw e;
		}
	}

	bool BaseSocket::SendOverSocket(const IPEndPoint& EndPoint, const BufferStream& Buffer)
	{
		try
		{
			if (m_PacketLossSimulation == 0 || Constants::Random.NextDouble() > m_PacketLossSimulation)
				SocketUtilities::SendTo(m_Socket, EndPoint, Buffer.GetBuffer(), Buffer.GetSize());

			m_Statistics.AddBandwidthOut(Buffer.GetSize());

			return true;
		}
		catch (PlatformNetwork::SocketException e)
		{
			if (e.GetError() == PlatformNetwork::Errors::ConnectionReset ||
				e.GetError() == PlatformNetwork::Errors::ConnectionAborted ||
				e.GetError() == PlatformNetwork::Errors::ConnectionRefused)
			{
				return false;
			}

			throw e;
		}
	}

	void BaseSocket::AddEvent(EventBase* Event)
	{
		WAIT_FOR_BOOL(m_EventsLock);

		m_Events.push_back(Event);
	}

	void BaseSocket::AddSendCommand(SendCommand* Command)
	{
		WAIT_FOR_BOOL(m_SendCommandsLock);

		m_SendCommands.push_back(Command);
	}

	void BaseSocket::HandleDisconnection(Socket Socket)
	{
		//Internationally Left Blank
	}

	void BaseSocket::ReceiverWorker(void)
	{
		while (true)
		{
			this_thread::sleep_for(chrono::milliseconds(1));

			Receive();
		}
	}

	void BaseSocket::SendWorker(void)
	{
		while (true)
		{
			this_thread::sleep_for(chrono::milliseconds(1));

			HandleSendCommands();
		}
	}

	void BaseSocket::HandleSendCommands(void)
	{
		WAIT_FOR_BOOL(m_SendCommandsLock);

		for (int i = 0; i < m_SendCommands.size(); ++i)
		{
			SendCommand* command = m_SendCommands[i];

			if (GetTimestamp() < command->GetSendTime() + (m_LatencySimulation / 1000.0F))
				continue;

			if (!HandleSendCommand(m_SendCommands[i]))
				continue;

			m_SendCommands.erase(m_SendCommands.begin() + i);

			i--;

			delete command;
		}
	}
}